db/schema_applier.cc

```diff
@@ -845,6 +841,19 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
     }
 }
 
+future<> schema_applier::lock_tables_metadata() {
+    // Adding and dropping tables, or changing tablet metadata, uses this
+    // locking mechanism to prevent changes to tables_metadata during preemptive
+    // iteration over it (e.g., tables_metadata::for_each_table_gently).
+    // However, we can only acquire the (write) lock after preparing all
+    // entities for the pending schema change that need to iterate over tables_metadata;
+    // otherwise, such iteration would deadlock.
+    auto& db = _proxy.local().get_db();
+    _metadata_locks = std::make_unique<
+            replica::tables_metadata_lock_on_all_shards>(
+            co_await replica::database::lock_tables_metadata(db));
```

Comment:
Why not take it in commit() on each shard to reduce the scope of the lock and thus impact on concurrent operations?